File: pe-core/src/main/java/org/saar/core/common/normalmap/NormalMappedMeshWriter.java
package org.saar.core.common.normalmap;

import org.saar.core.mesh.writers.ElementsMeshWriter;

public class NormalMappedMeshWriter implements ElementsMeshWriter<NormalMappedVertex> {

    private final NormalMappedMeshPrototype prototype;

    public NormalMappedMeshWriter(NormalMappedMeshPrototype prototype) {
        this.prototype = prototype;
    }

    @Override
    public void writeVertex(NormalMappedVertex vertex) {
        this.prototype.getPositionBuffer().getWriter().write3f(vertex.getPosition3f());
        this.prototype.getUvCoordBuffer().getWriter().write2f(vertex.getUvCoord2f());
        this.prototype.getNormalBuffer().getWriter().write3f(vertex.getNormal3f());
        this.prototype.getTangentBuffer().getWriter().write3f(vertex.getTangent3f());
        this.prototype.getBiTangentBuffer().getWriter().write3f(vertex.getBiTangent3f());
    }

    @Override
    public void writeIndex(int index) {
        this.prototype.getIndexBuffer().getWriter().writeInt(index);
    }

}

File: pe-core/src/main/java/org/saar/core/common/normalmap/NormalMappedMeshWriter.kt
package org.saar.core.common.normalmap

import org.saar.core.mesh.builder.IndexedMeshWriter
import org.saar.core.mesh.builder.VertexMeshWriter
import org.saar.lwjgl.util.DataWriter

class NormalMappedMeshWriter(
    private val positionWriter: DataWriter,
    private val uvCoordWriter: DataWriter,
    private val normalWriter: DataWriter,
    private val tangentWriter: DataWriter,
    private val biTangentWriter: DataWriter,
    private val indexWriter: DataWriter,
) : VertexMeshWriter<NormalMappedVertex>, IndexedMeshWriter {

    override fun writeVertex(vertex: NormalMappedVertex) {
        this.positionWriter.write3f(vertex.position3f)
        this.uvCoordWriter.write2f(vertex.uvCoord2f)
        this.normalWriter.write3f(vertex.normal3f)
        this.tangentWriter.write3f(vertex.tangent3f)
        this.biTangentWriter.write3f(vertex.biTangent3f)
    }

    override fun writeIndex(index: Int) = this.indexWriter.writeInt(index)
}